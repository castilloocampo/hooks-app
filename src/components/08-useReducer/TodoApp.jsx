import { useReducer } from 'react';

import './styles.css'

import todoReducer from './todoReducer';
import useForm from '../../hooks/useForm';
import { useEffect } from 'react';
import TodoList from './TodoList';


const initialState = [];

const initialFormState = {
  description: ""
};

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
}

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, initialState, init);
  const [{ description }, handleInputChange, resetForm] = useForm(initialFormState);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  useEffect(() => {

  }, [])

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim() !== "") {
      const newTodo = {
        id: new Date().getTime(),
        description,
        done: false
      }
      dispatch({ type: "ADD_TODO", payload: newTodo });
      resetForm();
    }
  }

  const handleDeleteTODO = (id) => {
    dispatch({ type: "DELETE_TODO", payload: id });
  }

  const handleToggleTODO = (id) => {
    dispatch({ type: "TOGGLE_TODO", payload: id });
  }

  return (
    <div>
      <h1>TodoApp â†’	 ( {todos.length} )</h1>
      <hr />

      <div className='flex'>
        <div className='w-3/5'>
          <TodoList todos={todos} onHandleToggleTODO={handleToggleTODO} onHandleDeleteTODO={handleDeleteTODO} />
        </div>
        <div>
          <h4>Agregar TODO</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              name="description"
              placeholder="Aprender..."
              autoComplete="off"
              value={description}
              onChange={handleInputChange} />

            <button
              type='submit'
              className='btn btn-outline-primary w-full' >
              Agregar
            </button>
          </form>
        </div>

      </div>



    </div>
  );
}

export default TodoApp;
