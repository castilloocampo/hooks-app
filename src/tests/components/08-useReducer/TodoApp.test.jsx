
import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';
import TodoApp from '../../../components/08-useReducer/TodoApp';

describe('TodoApp component', () => {
  let { container } = render(<TodoApp />);

  // jest.spyOn(Object.getPrototypeOf(localStorage), 'setItem')
  // Object.setPrototypeOf(localStorage.setItem, jest.fn());
  // let setItemJest = localStorage.setItem;

  beforeEach(() => {
    // localStorage.clear();
    // setItemJest = null;
    cleanup();
    const tempRender = render(<TodoApp />);
    container = tempRender.container;
    // jest.spyOn(Object.getPrototypeOf(localStorage), 'setItem')
    // Object.setPrototypeOf(localStorage.setItem, jest.fn());
    // setItemJest = localStorage.setItem;
    Storage.prototype.setItem = jest.fn();
  });


  test('should show correctly', () => {
    expect(container).toMatchSnapshot();
    // expect(setItemJest).toHaveBeenCalledTimes(1);
  });

  test('should add todo from this component', async () => {
    // expect(setItemJest).toHaveBeenCalledTimes(1);
    const value = "Learn something";
    const inputTodo = screen.queryByRole('textbox');
    const formTodo = screen.getByTestId('form-todoadd');
    fireEvent.change(inputTodo, { target: { value } });
    fireEvent.submit(formTodo, { preventDefault: jest.fn() });
    await waitFor(() => {
      expect(screen.queryByTestId('p-list-item')).toHaveTextContent(`1. ${value}`);
    });
    fireEvent.change(inputTodo, { target: { value: 'Learn something more' } });
    fireEvent.submit(formTodo, { preventDefault: jest.fn() });
    await waitFor(() => {
      expect(screen.queryAllByTestId('p-list-item').length).toBe(2);
    });
    // expect(setItemJest).toHaveBeenCalledTimes(3);
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    expect(container).toMatchSnapshot();
  });

  test('should eliminate todo ', () => {
    const inputTodo = screen.queryByRole('textbox');
    const formTodo = screen.getByTestId('form-todoadd');
    for (let index = 0; index < 5; index++) {
      const value = `Learn something ${index + 1}`;
      fireEvent.change(inputTodo, { target: { value } });
      fireEvent.submit(formTodo, { preventDefault: jest.fn() });
    }
    expect(screen.queryAllByTestId('p-list-item').length).toBe(5);
    const btnDelete = screen.getAllByTestId('delete-button')[0];
    fireEvent.click(btnDelete);
    expect(screen.queryAllByTestId('p-list-item').length).toBe(4);
    // });

  

  });

})
