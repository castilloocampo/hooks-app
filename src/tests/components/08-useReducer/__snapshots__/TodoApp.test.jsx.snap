// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp component should add todo from this component 1`] = `
<div>
  <div>
    <h1>
      TodoApp →  ( 
      2
       )
    </h1>
    <hr />
    <div
      class="flex"
    >
      <div
        class="w-3/5"
      >
        <ul>
          <li
            class="relative block py-3 px-6 -mb-px "
            id="list-group-item"
          >
            <p
              class="text-center false"
              data-testid="p-list-item"
            >
              1
              . 
              Learn something
            </p>
            <button
              class="btn btn-danger"
              data-testid="delete-button"
            >
              Borrar
            </button>
          </li>
          <li
            class="relative block py-3 px-6 -mb-px "
            id="list-group-item"
          >
            <p
              class="text-center false"
              data-testid="p-list-item"
            >
              2
              . 
              Learn something more
            </p>
            <button
              class="btn btn-danger"
              data-testid="delete-button"
            >
              Borrar
            </button>
          </li>
        </ul>
      </div>
      <div
        class="w-2/5"
      >
        <h4>
          Agregar TODO
        </h4>
        <hr />
        <form
          data-testid="form-todoadd"
        >
          <input
            autocomplete="off"
            data-testid="input-todoadd"
            name="description"
            placeholder="Aprender..."
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary w-full"
            data-testid="button-todoadd"
            type="submit"
          >
            Agregar
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoApp component should show correctly 1`] = `
<div>
  <div>
    <h1>
      TodoApp →  ( 
      0
       )
    </h1>
    <hr />
    <div
      class="flex"
    >
      <div
        class="w-3/5"
      >
        <ul />
      </div>
      <div
        class="w-2/5"
      >
        <h4>
          Agregar TODO
        </h4>
        <hr />
        <form
          data-testid="form-todoadd"
        >
          <input
            autocomplete="off"
            data-testid="input-todoadd"
            name="description"
            placeholder="Aprender..."
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary w-full"
            data-testid="button-todoadd"
            type="submit"
          >
            Agregar
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
